autogen/agentchat/conversable_agent.py

```diff
@@ -983,6 +983,11 @@ def clear_history(self, recipient: Optional[Agent] = None, nr_messages_to_preser
         if recipient is None:
             if nr_messages_to_preserve:
                 for key in self._oai_messages:
+                    # if breaking history between function call and function response, save function call message
+                    # additionally, otherwise openai will return error
+                    first_msg_to_save = self._oai_messages[key][-nr_messages_to_preserve]
+                    if "tool_responses" in first_msg_to_save:
+                        nr_messages_to_preserve += 1
```

Comment:
This will change `nr_messages_to_preserve` for every agent. Do you intend to change it for the agent corresponding to the current key only?
Also, it'll be useful to add a comment that an assumption is made about the previous message in the list.

Reply:
@sonichi `nr_messages_to_preserve` variable works inside of `clear_history()` function, which called for every agent separately. Then, it should change `nr_messages_to_preserve` for curret agent only. Or I'm understanding something bad?

Added commentary for it.

Reply:
The for loop in line 1025 will check each key. Each key corresponds to one agent.

Reply:
I think it isn't. for loop in line 1025 used to convert default_dict with history to list representation. That loop always has only one iteration. As I'm testing it now, before entering the loop `self._oai_messages` looks like `defaultdict(<class 'list'>, {<autogen.agentchat.groupchat.GroupChatManager object at 0x000001FA75C3F250>: [{'content': 'message here'}, {'content': 'other message here'}]}]`, but loop allows to convert it to more understandable form, where `self._oai_messages[key]` is just `[{'content': 'message here'}, {'content': 'other message here'}]`. I know, it's not really elegant to use loop just to convert variable, maybe there is better way, but I didn't found it, I have no big expertise with default_dicts.

Reply:
In general it's not guaranteed to have only one key. When an agent is involved in multiple conversations, it can have multiple keys. We should reset the counter for every key.

Reply:
@sonichi ok, improved it